flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala

```diff
@@ -298,11 +333,34 @@ case class Join(
     val resolvedJoin = super.validate(tableEnv).asInstanceOf[Join]
     if (!resolvedJoin.condition.forall(_.resultType == BOOLEAN_TYPE_INFO)) {
       failValidation(s"filter expression ${resolvedJoin.condition} is not a boolean")
-    } else if (!ambiguousName.isEmpty) {
+    } else if (ambiguousName.nonEmpty) {
       failValidation(s"join relations with ambiguous names: ${ambiguousName.mkString(", ")}")
     }
+
+    resolvedJoin.condition.foreach(testJoinCondition(_))
     resolvedJoin
   }
+
+  private def testJoinCondition(expression: Expression): Unit = {
+    def checkIfJoinCondition(exp : Expression) = if (exp.children.exists(!_.isInstanceOf[JoinFieldReference])) {
```

Comment:
The condition here should be: all (of the two) children are `JoinFieldReference` and one is on left and the other on the right input (that's where the input flag in `JoinFieldReference` is helpful).

I would also change the type of `exp` to `BinaryComparison`